doc/melange.rst

```diff
@@ -0,0 +1,298 @@
+  listed dependencies in ``libraries``. See the documentation for
+  ``root_module`` in the :ref:`library` stanza.
+
+- ``(allow_overlapping_dependencies)`` is the same as the corresponding field of
+  :ref:`library`.
+
+Libraries
+=========
+
+Melange is compatible with Dune libraries that support it. When adding support
+for Melange in libraries, two fields are involved:
+
+- ``(modes melange)``: it is required to add ``melange`` to the library
+  ``modes`` field.
+
+  - If the library has no ``modes`` field defined yet, one can use the
+    :doc:`reference/ordered-set-language` to do so: ``(modes :standard melange)``
+
+  - If the library already had some value for the ``modes`` field, ``melange``
+    mode can be appended to the list of modes, for example:
+    ``(modes byte native melange)``
+
+- ``(melange.runtime_deps <deps>)``: optionally, define any runtime dependencies
+  using ``melange.runtime_deps``. This field is analog to the ``runtime_deps``
+  field used in ``melange.emit`` stanzas. An example:
+  ``(melange.runtime_deps ./path/to/file.css (glob_file_rec ./fonts/*))``.
+
+Recommended Practices
+=====================
+
+Keep Bundles Small by Reducing the Number of ``melange.emit`` Stanzas
+---------------------------------------------------------------------
+
+It is recommended to define a single ``melange.emit`` stanza for each project,
+or if that is not possible, at least minimize the number of ``melange.emit``
+stanzas defined.
+
+The reason is that using multiple ``melange.emit`` stanzas will lead to multiple
+copies of the JavaScript files generated for library code. For example, if
+there are two ``melange.emit`` stanzas defined as follows:
+
+.. code:: dune
+
+  (melange.emit
+   (target app1)
+   (libraries foo))
+
+  (melange.emit
+   (target app2)
+   (libraries foo))
+   
+The JavaScript artifacts for library ``foo`` will end up duplicated in the
+``_build`` folder, both in ``_build/default/app1`` and ``_build/default/app2``.
+
+This can have unexpected impact on bundle size when using tools like Webpack or
+Esbuild, as these tools will not be able to see shared library code as such,
+as it would be replicated across the paths of the different stanzas
+``target`` folders.
+
+
+Faster Builds with ``subdir`` and ``dirs`` Stanzas
+--------------------------------------------------
+
+Melange libraries might be installed from the ``npm`` package repository,
+together with other JavaScript packages. To avoid having Dune inspect
+unnecessary folders in ``node_modules``, it is recommended to explicitly
+include those folders that are relevant for Melange builds.
+
+This can be accomplished by combining :ref:`subdir` and :ref:`dune-subdirs`
+stanzas in a ``dune`` file and by co-locating this file together with the
+``node_modules`` folder. The :ref:`dune-vendored_dirs` stanza can be used as
+well to avoid warnings in Melange libraries during the application build.
```

Comment:
```suggestion
This can be accomplished by combining :ref:`subdir` and :ref:`dune-subdirs`
stanzas in a ``dune`` file next to the ``node_modules`` folder.  The 
:ref:`dune-vendored_dirs` stanza can be used to avoid warnings in Melange
libraries during the application build. The :ref:`dune-data_only_dirs` stanza
can be useful as well if you need to override the build rules in one of the packages.
```